tools/src/post-mortem/Statistics/HydraMetrics.hs

```diff
@@ -0,0 +1,58 @@
+module Statistics.HydraMetrics
+    ( findBlockChainState
+    , ChainState(..)
+    ) where
+
+import           Control.Foldl   (Fold (..))
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as M
+import           JSONLog         (IndexedJLTimedEvent, IndexedJLTimedEvent(..))
+import           Pos.Util.JsonLog(JLBlock, JLBlock(..),  JLEvent(..))
+import           Universum
+
+type JLSlotId = (Word64, Word16)
+type BlockId = Text
+
+data ChainStateInternal = ChainStateInternal
+  { topMostSlot :: JLSlotId
+  , topMostBlock :: BlockId
+  , blocks :: Map BlockId BlockWrapper
+  } deriving Show
+
+data ChainState = ChainState
+  { internal :: ChainStateInternal
+  , chainLength :: Integer
+  , expectedLength :: Integer
+  } deriving Show
+
+data BlockWrapper = BlockWrapper
+  { jlBlock :: JLBlock
+  , parent :: Maybe BlockWrapper
+  } deriving Show
+
+findBlockChainState :: Fold IndexedJLTimedEvent ChainState
+findBlockChainState = Fold combine (ChainStateInternal (0,0) "error" mempty) finish
+  where
+    finish :: ChainStateInternal -> ChainState
+    finish internal = ChainState internal chainLength (((fromIntegral slot) + 1) + ((fromIntegral epoch) * epochLength))
+      where
+        chainLength = measureLength $ M.lookup (topMostBlock internal) (blocks internal)
+        (epoch, slot) = topMostSlot internal
+        epochLength = 1080 -- TODO, look it up somehow
+    measureLength :: Maybe BlockWrapper -> Integer
+    measureLength (Just block) = 1 + measureLength (parent block)
```

Comment:
For very big chains, you might want to go with a tail-recursive version of this function, where you would accumulate the counter in an aux variable and force the sum with a `BangPattern` to avoid building too many thunks.

Reply:
yeah, i can see that helping, though right now the chain doesnt go over 30 blocks, i'll add a todo